cql3/expr/expression.cc

```diff
@@ -931,6 +931,61 @@ std::ostream& operator<<(std::ostream& s, oper_t op) {
     __builtin_unreachable();
 }
 
+extern std::vector<expression> extract_column_restrictions(const expression& expr, const column_definition& column) {
+    struct visitor {
+        std::vector<expression> restrictions;
+        const column_definition& column;
+        const binary_operator* current_binary_operator;
+
+        void operator()(bool b) {
+            restrictions.emplace_back(b);
+        }
+
+        void operator()(const conjunction& conj) {
+            for (const expression& child : conj.children) {
+                std::visit(*this, child);
+            }
+        }
+
+        void operator()(const binary_operator& oper) {
+            if (current_binary_operator != nullptr) {
+                on_internal_error(expr_logger, "extract_column_restrictions: nested binary operators are not supported");
+            }
+
+            current_binary_operator = &oper;
+            std::visit(*this, *oper.lhs);
+            current_binary_operator = nullptr;
+        }
+
+        void operator()(const column_value& cv) {
+            if (*cv.col == column && current_binary_operator != nullptr) {
+                restrictions.emplace_back(*current_binary_operator);
+            }
+        }
+
+        void operator()(const column_value_tuple& tuple) {
+            for (const column_value& cv : tuple.elements) {
+                if (*cv.col == column && current_binary_operator != nullptr) {
+                    restrictions.emplace_back(*current_binary_operator);
```

Comment:
Should you extract the column from the tuple?

For example, if the relation was `(x, y) = (1, 2)`, and you're looking for `x`, should you return `x = 1`?

Reply:
We could do this for equality, but `(x, y) < (1, 2)` wouldn't work this way.
Maybe it would be better to extract relations as is and let multi column code handle the rest.

Reply:
I imaging that now it doesn't work either for tuple equality or tuple inequality. If that's the case let's skip it, and later consider whether to special-case tuple equality.

Reply:
I see it's still here. Isn't it broken?

Reply:
I think that depends on the caller.  It's not wrong, in principle, for this function to return `(c1,c2)>?` as one of "all binary operators which have the given column on their left hand side" (quoted from the functions' blurb).

The caller currently assumes all single-column LHS and therefore invokes get_single_column_clustering_bounds on the result.  When #9085 is fixed, this assumption will be correct.  You're right, of course, that the caller could verify the assumption and call on_internal_error if it's violated.

Reply:
> It's not wrong, in principle, for this function to return `(c1,c2)>?` as one of "all binary operators which have the given column on their left hand side" (quoted from the functions' blurb).

I changed my mind (see other comment in this review).  Please skip multi-column binops here and update the blurb accordingly.